include/SFML/System/UniqueResource.inl

```diff
@@ -0,0 +1,106 @@
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+UniqueResource<T, Deleter>::UniqueResource() : m_resource(), m_deleter(), m_delete(false)
+{
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+UniqueResource<T, Deleter>::UniqueResource(T&& resource, Deleter&& deleter) :
+m_resource(std::move(resource)),
+m_deleter(std::move(deleter)),
+m_delete(true)
+{
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+UniqueResource<T, Deleter>::~UniqueResource()
+{
+    reset();
+}
+
+template <typename T, typename Deleter>
+UniqueResource<T, Deleter>::UniqueResource(UniqueResource&& other) noexcept :
+m_resource(std::move(other.m_resource)),
+m_deleter(std::move(other.m_deleter)),
+m_delete(std::move(other.m_delete))
+{
+    other.release();
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+UniqueResource<T, Deleter>& UniqueResource<T, Deleter>::operator=(UniqueResource&& other) noexcept
+{
+    reset();
+    m_deleter  = std::move(other.m_deleter);
+    m_resource = std::move(other.m_resource);
+    m_delete   = std::move(other.m_delete);
+    other.release();
+    return *this;
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+void UniqueResource<T, Deleter>::release()
+{
+    m_delete = false;
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+void UniqueResource<T, Deleter>::reset()
+{
+    if (m_delete)
+        m_deleter(m_resource);
+    m_delete = false;
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+void UniqueResource<T, Deleter>::reset(T&& resource)
+{
+    reset();
+    m_resource = std::move(resource);
+    m_delete   = true;
+}
+
+
+////////////////////////////////////////////////////////////
+template <typename T, typename Deleter>
+const T& UniqueResource<T, Deleter>::get() const
+{
+    return m_resource;
+}
```

Comment:
This could have an assertion to check that the resource is populated.

Reply:
Are there any circumstances under which you'd want to access this handle without accessing the resource it points to? If so then we can't `assert` here but I'm struggling to come up with such an example.